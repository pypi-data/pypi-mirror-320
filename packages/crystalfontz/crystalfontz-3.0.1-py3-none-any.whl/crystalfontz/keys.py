from dataclasses import asdict, dataclass
from enum import Enum
from typing import Any, Dict, List, Type

try:
    from typing import Self
except ImportError:
    Self = Any

KeyPress = int

KP_UP: KeyPress = 0x01
KP_ENTER: KeyPress = 0x02
KP_EXIT: KeyPress = 0x04
KP_LEFT: KeyPress = 0x08
KP_RIGHT: KeyPress = 0x10
KP_DOWN: KeyPress = 0x20


@dataclass
class KeyState:
    """
    A key's state.

    Attributes:
        pressed (bool): When True, the key is currently pressed.
        pressed_since (bool): When True, the key has been pressed since the last poll.
        released_since (bool): When True, the key has been released since the last
                               poll.
    """

    pressed: bool
    pressed_since: bool
    released_since: bool


@dataclass
class KeyStates:
    """
    The state of all keys.

    Attributes:
        up: The state of the "up" key.
        enter: The state of the "enter" key.
        exit: The state of the "exit" key.
        left: The state of the "left" key.
        right: The state of the "right" key.
        down: The state of the "down" key.
    """

    up: KeyState
    enter: KeyState
    exit: KeyState
    left: KeyState
    right: KeyState
    down: KeyState

    @classmethod
    def from_bytes(cls: Type[Self], state: bytes) -> Self:
        pressed = state[0]
        pressed_since = state[1]
        released_since = state[2]

        return cls(
            up=KeyState(
                pressed=bool(pressed & KP_UP),
                pressed_since=bool(pressed_since & KP_UP),
                released_since=bool(released_since & KP_UP),
            ),
            enter=KeyState(
                pressed=bool(pressed & KP_ENTER),
                pressed_since=bool(pressed_since & KP_ENTER),
                released_since=bool(released_since & KP_ENTER),
            ),
            exit=KeyState(
                pressed=bool(pressed & KP_EXIT),
                pressed_since=bool(pressed_since & KP_EXIT),
                released_since=bool(released_since & KP_EXIT),
            ),
            left=KeyState(
                pressed=bool(pressed & KP_LEFT),
                pressed_since=bool(pressed_since & KP_LEFT),
                released_since=bool(released_since & KP_LEFT),
            ),
            right=KeyState(
                pressed=bool(pressed & KP_RIGHT),
                pressed_since=bool(pressed_since & KP_RIGHT),
                released_since=bool(released_since & KP_RIGHT),
            ),
            down=KeyState(
                pressed=bool(pressed & KP_DOWN),
                pressed_since=bool(pressed_since & KP_DOWN),
                released_since=bool(released_since & KP_DOWN),
            ),
        )

    def as_dict(self: Self) -> Dict[str, Dict[str, bool]]:
        return {key: state for key, state in asdict(self).items()}


class KeyActivity(Enum):
    """
    A key activity. This is either a "press" event or a "release" event for a
    given key.
    """

    KEY_UP_PRESS = 1
    KEY_DOWN_PRESS = 2
    KEY_LEFT_PRESS = 3
    KEY_RIGHT_PRESS = 4
    KEY_ENTER_PRESS = 5
    KEY_EXIT_PRESS = 6
    KEY_UP_RELEASE = 7
    KEY_DOWN_RELEASE = 8
    KEY_LEFT_RELEASE = 9
    KEY_RIGHT_RELEASE = 10
    KEY_ENTER_RELEASE = 11
    KEY_EXIT_RELEASE = 12

    @classmethod
    def from_bytes(cls: Type[Self], activity: bytes) -> "KeyActivity":
        return KEY_ACTIVITIES[activity[0] - 1]


KEY_ACTIVITIES: List[KeyActivity] = [
    KeyActivity.KEY_UP_PRESS,
    KeyActivity.KEY_DOWN_PRESS,
    KeyActivity.KEY_LEFT_PRESS,
    KeyActivity.KEY_RIGHT_PRESS,
    KeyActivity.KEY_ENTER_PRESS,
    KeyActivity.KEY_EXIT_PRESS,
    KeyActivity.KEY_UP_RELEASE,
    KeyActivity.KEY_DOWN_RELEASE,
    KeyActivity.KEY_LEFT_RELEASE,
    KeyActivity.KEY_RIGHT_RELEASE,
    KeyActivity.KEY_ENTER_RELEASE,
    KeyActivity.KEY_EXIT_RELEASE,
]
