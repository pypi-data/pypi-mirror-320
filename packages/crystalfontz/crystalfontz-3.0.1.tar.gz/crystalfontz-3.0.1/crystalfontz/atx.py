from dataclasses import dataclass
from enum import Enum
from typing import Any, Optional, Set, Type

try:
    from typing import Self
except ImportError:
    Self = Any


AUTO_POLARITY = 0x01
RESET_INVERT = 0x02
POWER_INVERT = 0x04


class AtxPowerSwitchFunction(Enum):
    """
    An ATX power switch function.

    Refer to your device's datasheet for the effects of each of these functions.
    """

    LCD_OFF_IF_HOST_IS_OFF = 0x10
    KEYPAD_RESET = 0x20
    KEYPAD_POWER_ON = 0x40
    KEYPAD_POWER_OFF = 0x80


@dataclass
class AtxPowerSwitchFunctionalitySettings:
    functions: Set[AtxPowerSwitchFunction]
    auto_polarity: bool = True
    reset_invert: bool = False
    power_invert: bool = False
    power_pulse_length_seconds: Optional[float] = None

    """
    Settings for command 28 (0x1C): Set ATX Power Switch Functionality.

    Parameters:
        functions: A set of enabled power switch functions.
        auto_polarity: When True, automatically detects polarity for reset and/or
                       power (recommended)
        reset_invert: When True, the reset pin drives high instead of low
        power_invert: When True, the power pin drives high instead of low
        power_pulse_length_seconds:  Length of power on and off pulses in seconds.
                                     When set to 8 seconds or higher, asserts power
                                     control line until host power state changes.

    """

    @classmethod
    def from_bytes(cls: Type[Self], settings: bytes) -> Self:
        functions: Set[AtxPowerSwitchFunction] = set()
        for function in AtxPowerSwitchFunction:
            if settings[0] & function.value:
                functions.add(function)
        auto_polarity: bool = bool(settings[0] & AUTO_POLARITY)
        power_pulse_length_seconds: Optional[float] = (
            settings[1] / 32 if len(settings) > 1 else None
        )

        return cls(
            functions=functions,
            auto_polarity=auto_polarity,
            power_pulse_length_seconds=power_pulse_length_seconds,
        )

    def to_bytes(self: Self) -> bytes:
        functions: int = 0
        for function in self.functions:
            functions ^= function.value
        if self.auto_polarity:
            functions ^= AUTO_POLARITY
        if self.reset_invert:
            functions ^= RESET_INVERT
        if self.power_invert:
            functions ^= POWER_INVERT

        packed: bytes = functions.to_bytes(length=1)

        if self.power_pulse_length_seconds is not None:
            pulse_length = int(self.power_pulse_length_seconds * 32)

            if pulse_length < 1:
                raise ValueError(f"Pulse length can not be less than {1/32}")

            packed += min(pulse_length, 255).to_bytes(length=1)

        return packed
