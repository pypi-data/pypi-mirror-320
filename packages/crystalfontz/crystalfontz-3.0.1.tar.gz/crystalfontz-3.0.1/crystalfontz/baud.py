from typing import Literal

BaudRate = Literal[19200] | Literal[115200]

SLOW_BAUD_RATE: BaudRate = 19200
FAST_BAUD_RATE: BaudRate = 115200
