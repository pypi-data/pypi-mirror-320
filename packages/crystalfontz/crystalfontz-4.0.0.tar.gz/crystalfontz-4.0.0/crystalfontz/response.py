from abc import ABC, abstractmethod
from dataclasses import asdict
import struct
import textwrap
from typing import Any, Callable, cast, Dict, Type, TypeVar

try:
    from typing import Self
except ImportError:
    Self = Any

from crystalfontz.error import (
    DecodeError,
    DeviceError,
    ResponseDecodeError,
    UnknownResponseError,
)
from crystalfontz.format import format_bytes, format_json_bytes
from crystalfontz.gpio import GpioSettings, GpioState
from crystalfontz.keys import KeyActivity, KeyStates
from crystalfontz.packet import Packet


def assert_len(n: int, data: bytes) -> None:
    if len(data) != n:
        raise DecodeError(f"Response expected to be {n} bytes, is {len(data)} bytes")


class Response(ABC):
    """
    A response received from the Crystalfontz LCD.

    To implement a new response type, subclass this class and implement the
    __init__ method.
    """

    @abstractmethod
    def __init__(self: Self, data: bytes) -> None:
        raise NotImplementedError("__init__")

    @classmethod
    def from_packet(cls: Type[Self], packet: Packet) -> "Response":
        code, data = packet
        if code in RESPONSE_CLASSES:
            res_cls = RESPONSE_CLASSES[code]
            try:
                return res_cls(data)
            except Exception as exc:
                raise ResponseDecodeError(res_cls, str(exc)) from exc

        if DeviceError.is_error_code(code):
            raise DeviceError(packet)

        raise UnknownResponseError(packet)


class RawResponse(Response):
    """
    A raw response. This class may be used with `client.expect` to capture
    an otherwise unsupported response type.
    """

    def __init__(self: Self, data: bytes) -> None:
        self.code: int = 0xFF
        self.data: bytes = data

    @classmethod
    def from_packet(cls: Type[Self], packet: Packet) -> Self:
        code, data = packet
        res = cls(data)
        res.code = code
        return res


class Ack(Response):
    def __init__(self: Self, data: bytes) -> None:
        assert_len(0, data)

    def __str__(self: Self) -> str:
        return "Ack()"


RESPONSE_CLASSES: Dict[int, Type[Response]] = {}

R = TypeVar("R", bound=Response)


def code(code: int) -> Callable[[Type[R]], Type[R]]:
    def decorator(cls: Type[R]) -> Type[R]:
        RESPONSE_CLASSES[code] = cast(Type[Response], cls)
        return cls

    return decorator


@code(0x40)
class Pong(Response):
    """
    Attributes:
        response (bytes): The data sent in the ping command.
    """

    def __init__(self: Self, data: bytes) -> None:
        self.response = data

    def __str__(self: Self) -> str:
        return f"Pong({self.response})"


@code(0x41)
class Versions(Response):
    """
    Attributes:
        model (str): The device model.
        hardware_rev (str): The device's hardware revision.
        firmware_rev (str): The device's firmware revision.
    """

    def __init__(self: Self, data: bytes) -> None:
        decoded = data.decode("ascii")
        model, versions = decoded.split(":")
        hw_rev, fw_rev = versions.split(",")

        self.model: str = model
        self.hardware_rev: str = hw_rev.strip()
        self.firmware_rev: str = fw_rev.strip()

    def __str__(self: Self) -> str:
        return (
            f"Versions(model={self.model}, hardware_rev={self.hardware_rev}, "
            f"firmware_rev={self.firmware_rev})"
        )

    def as_dict(self: Self) -> Dict[str, Any]:
        return dict(
            model=self.model,
            hardware_rev=self.hardware_rev,
            firmware_rev=self.firmware_rev,
        )

    def __repr__(self: Self) -> str:
        return f"{self.model}: {self.hardware_rev}, {self.firmware_rev}"


@code(0x42)
class UserFlashAreaWritten(Ack):
    def __str__(self: Self) -> str:
        return "UserFlashAreaWritten()"


@code(0x43)
class UserFlashAreaRead(Response):
    """
    Attributes:
        data (bytes): The data read from the user flash area.
    """

    def __init__(self: Self, data: bytes) -> None:
        self.data: bytes = data

    def __str__(self: Self) -> str:
        return f"UserFlashAreaRead({self.data})"


@code(0x44)
class BootStateStored(Ack):
    def __str__(self: Self) -> str:
        return "BootStateStored()"


@code(0x45)
class PowerResponse(Ack):
    def __str__(self: Self) -> str:
        return "PowerResponse()"


@code(0x46)
class ClearedScreen(Ack):
    def __str__(self: Self) -> str:
        return "ClearedScreen()"


@code(0x47)
class Line1Set(Ack):
    def __str__(self: Self) -> str:
        return "Line1Set()"


@code(0x48)
class Line2Set(Ack):
    def __str__(self: Self) -> str:
        return "Line2Set()"


@code(0x49)
class SpecialCharacterDataSet(Ack):
    def __str__(self: Self) -> str:
        return "SpecialCharacterDataSet()"


@code(0x4A)
class LcdMemory(Response):
    """
    Attributes:
        address (int): The address read from LCD memory.
        data (bytes): The data read from the address in LCD memory.
    """

    def __init__(self: Self, data: bytes) -> None:
        assert_len(9, data)
        self.address: int = data[0]
        self.data: bytes = data[1:]

    def __str__(self: Self) -> str:
        return f"LcdMemory(0x{self.address:02X}={self.data})"

    def as_dict(self: Self) -> Dict[str, Any]:
        return dict(address=self.address, data=format_json_bytes(self.data))

    def __repr__(self: Self) -> str:
        return f"0x{self.address:02X}: {format_bytes(self.data)}"


@code(0x4B)
class CursorPositionSet(Ack):
    def __str__(self: Self) -> str:
        return "CursorPositionSet()"


@code(0x4C)
class CursorStyleSet(Ack):
    def __str__(self: Self) -> str:
        return "CursorStyleSet()"


@code(0x4D)
class ContrastSet(Ack):
    def __str__(self: Self) -> str:
        return "ContrastSet()"


@code(0x4E)
class BacklightSet(Ack):
    def __str__(self: Self) -> str:
        return "BacklightSet()"


@code(0x52)
class DowDeviceInformation(Response):
    """
    Attributes:
        index (int): The DOW device index.
        rom_id (bytes): The ROM ID of the device.
    """

    def __init__(self: Self, data: bytes) -> None:
        self.index: int = data[0]
        self.rom_id: bytes = data[1:]

    def __str__(self: Self) -> str:
        return f"DowDeviceInformation({self.index}={self.rom_id})"

    def as_dict(self: Self) -> Dict[str, Any]:
        return dict(index=self.index, rom_id=format_json_bytes(self.rom_id))

    def __repr__(self: Self) -> str:
        return f"0x{self.index:02X}: {format_bytes(self.rom_id)}"


@code(0x53)
class TemperatureReportingSetUp(Ack):
    def __str__(self: Self) -> str:
        return "TemperatureReportingSetUp()"


@code(0x54)
class DowTransactionResult(Response):
    """
    Attributes:
        index (int): The DOW device index.
        data (bytes): Data read from the 1-wire bus.
        crc (int): The 1-wire CRC.
    """

    def __init__(self: Self, data: bytes) -> None:
        self.index = data[0]
        self.data = data[1:-1]
        self.crc = data[-1]

    def __str__(self: Self) -> str:
        return (
            f"DowTransactionResult(0x{self.index:02X}, "
            f"data={self.data}, crc={self.crc})"
        )

    def as_dict(self: Self) -> Dict[str, Any]:
        return dict(index=self.index, data=format_json_bytes(self.data), crc=self.crc)

    def __repr__(self: Self) -> str:
        repr_ = f"Transaction Result for Device {self.index}:\n"
        repr_ += f"  Data: {format_bytes(self.data)}\n"
        repr_ += f"  CRC: 0x{self.crc:02X}"
        return repr_


@code(0x55)
class LiveTemperatureDisplaySetUp(Ack):
    def __str__(self: Self) -> str:
        return "LiveTemperatureDisplaySetUp"


@code(0x56)
class CommandSentToLcdController(Ack):
    def __str__(self: Self) -> str:
        return "CommandSentToLcdController()"


@code(0x57)
class KeyReportingConfigured(Ack):
    def __str__(self: Self) -> str:
        return "KeyReportingConfigured()"


@code(0x58)
class KeypadPolled(Response):
    """
    Attributes:
        states (KeyStates): The keypad's key states.
    """

    def __init__(self: Self, data: bytes) -> None:
        assert_len(3, data)
        self.states = KeyStates.from_bytes(data)

    def __str__(self: Self) -> str:
        return f"KeypadPolled(states={self.states})"

    def as_dict(self: Self) -> Dict[str, Any]:
        return dict(states=asdict(self.states))

    def __repr__(self: Self) -> str:
        repr_ = "Keypad States:\n"
        repr_ += textwrap.indent(repr(self.states), "  ")

        return repr_


@code(0x5C)
class AtxPowerSwitchFunctionalitySet(Ack):
    def __str__(self: Self) -> str:
        return "AtxPowerFunctionalitySet()"


@code(0x5D)
class WatchdogConfigured(Ack):
    def __str__(self: Self) -> str:
        return "WatchdogConfigured()"


@code(0x5E)
class StatusRead(Response):
    """
    A raw status response. This status is parsed based on the device.
    """

    def __init__(self: Self, data: bytes) -> None:
        self.data: bytes = data

    def __str__(self: Self) -> str:
        return f"StatusRead({self.data})"


@code(0x5F)
class DataSent(Ack):
    def __str__(self: Self) -> str:
        return "DataSent()"


@code(0x61)
class BaudRateSet(Ack):
    def __str__(self: Self) -> str:
        return "BaudRateSet()"


@code(0x62)
class GpioSet(Ack):
    def __str__(self: Self) -> str:
        return "GpioSet()"


@code(0x63)
class GpioRead(Response):
    """
    Attributes:
        index (int): The index of the GPIO pin.
        state (GpioState): Pin state & changes since last poll.
        requested_level (int): Requested Pin level/PWM level.
        settings (GpioSettings): Pin function select and drive mode.
    """

    def __init__(self: Self, data: bytes) -> None:
        self.index: int = data[0]
        self.state: GpioState = GpioState.from_byte(data[1])
        self.requested_level: int = data[2]
        self.settings: GpioSettings = GpioSettings.from_byte(data[3])

    def __str__(self: Self) -> str:
        return (
            f"GpioRead({self.index}, state={self.state}, "
            f"requested_level={self.requested_level}, settings={self.settings}"
        )

    def as_dict(self: Self) -> Dict[str, Any]:
        return dict(
            index=self.index,
            state=asdict(self.state),
            requested_level=self.requested_level,
            settings=self.settings.as_dict(),
        )

    def __repr__(self: Self) -> str:
        repr_ = f"GPIO pin {self.index}:\n"
        repr_ += f"  Requested Level: {self.requested_level}\n"
        repr_ += "  Settings:\n"
        repr_ += textwrap.indent(repr(self.settings), "    ")
        return repr_


@code(0x80)
class KeyActivityReport(Response):
    """
    A key activity report from the Crystalfontz LCD.

    Attributes:
        activity (KeyActivity): The reported key activity.
    """

    def __init__(self: Self, data: bytes) -> None:
        assert_len(1, data)

        self.activity: KeyActivity = KeyActivity.from_bytes(data)

    def __str__(self: Self) -> str:
        return f"KeyActivityReport({self.activity.name})"

    def __repr__(self: Self) -> str:
        return f"{self.__class__.__name__}\t{self.activity.name}"

    def as_dict(self: Self) -> Dict[str, Any]:
        return dict(type=self.__class__.__name__, activity=self.activity.name)


@code(0x82)
class TemperatureReport(Response):
    """
    A temperature sensor report from the Crystalfontz LCD.

    Attributes:
        index (int): The index of the temperature sensor.
        celsius (float): The temperature in celsius.
        fahrenheit (float): The temperature in fahrenheit.
    """

    def __init__(self: Self, data: bytes) -> None:
        assert_len(4, data)

        self.index: int = data[0]
        value = struct.unpack(">H", data[1:3])[0]
        dow_crc_status = data[3]

        if dow_crc_status == 0:
            raise DecodeError("Bad CRC from temperature sensor")

        self.celsius: float = value / 16.0
        self.fahrenheit: float = (9 / 5 * self.celsius) + 32.0

    def __str__(self: Self) -> str:
        return (
            f"TemperatureReport({self.index}, celsius={self.celsius}, "
            f"fahrenheit={self.fahrenheit})"
        )

    def __repr__(self: Self) -> str:
        return f"{self.__class__.__name__}\t{self.celsius}\t{self.fahrenheit}"

    def as_dict(self: Self) -> Dict[str, Any]:
        return dict(
            type=self.__class__.__name__,
            celsius=self.celsius,
            fahrenheit=self.fahrenheit,
        )
