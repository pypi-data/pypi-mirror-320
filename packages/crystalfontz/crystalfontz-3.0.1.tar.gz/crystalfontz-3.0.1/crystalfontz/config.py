from dataclasses import asdict, dataclass, field, fields, replace
import os
import os.path
from typing import Any, cast, Dict, Optional, Type

try:
    from typing import Self
except ImportError:
    Self = Any

from appdirs import user_config_dir
import yaml

from crystalfontz.baud import BaudRate, SLOW_BAUD_RATE

try:
    from yaml import CDumper as Dumper
    from yaml import CLoader as Loader
except ImportError:
    from yaml import Dumper, Loader

"""
Configuration management for the Crystalfontz CLI.
"""

APP_NAME = "crystalfontz"

GLOBAL_FILE = f"/etc/{APP_NAME}.yaml"


def default_file() -> str:
    """
    Get the default file path for the crystalfontz CLI.
    """

    return os.path.join(user_config_dir(APP_NAME), f"{APP_NAME}.yaml")


DEFAULT_PORT = "/dev/ttyUSB0"


def _metadata(env_var: Optional[str] = None) -> Dict[str, Any]:
    return dict(env_var=env_var)


def _from_environment() -> Dict[str, Any]:
    env: Dict[str, Any] = dict()
    for f in fields(Config):
        if f.metadata and "env_var" in f.metadata:
            if f.metadata["env_var"] in os.environ:
                var: Any = os.environ[f.metadata["env_var"]]
                env[f.name] = cast(Any, f.type)(var)
    return env


@dataclass
class Config:
    """
    A config for the crystalfontz CLI.
    """

    port: str = field(
        default=DEFAULT_PORT, metadata=_metadata(env_var="CRYSTALFONTZ_PORT")
    )
    model: str = field(default="CFA533", metadata=_metadata("CRYSTALFONTZ_MODEL"))
    hardware_rev: Optional[str] = field(
        default=None, metadata=_metadata(env_var="CRYSTALFONTZ_HARDWARE_REV")
    )
    firmware_rev: Optional[str] = field(
        default=None, metadata=_metadata(env_var="CRYSTALFONTZ_FIRMWARE_REV")
    )
    baud_rate: BaudRate = field(
        default=SLOW_BAUD_RATE, metadata=_metadata(env_var="CRYSTALFONTZ_BAUD_RATE")
    )
    timeout: Optional[float] = field(
        default=None, metadata=_metadata(env_var="CRYSTALFONTZ_TIMEOUT")
    )
    retry_times: Optional[int] = field(
        default=None, metadata=_metadata(env_var="CRYSTALFONTZ_RETRY_TIMES")
    )
    file: Optional[str] = None

    @classmethod
    def from_environment(cls: Type[Self]) -> Self:
        """
        Load a config from the environment.
        """

        return cls(**_from_environment())

    @classmethod
    def from_file(
        cls: Type[Self],
        file: Optional[str] = None,
        load_environment: bool = False,
        create_file: bool = False,
    ) -> Self:
        """
        Load a config from a file.
        """

        _file: str = file or os.environ.get("CRYSTALFONTZ_CONFIG", default_file())

        found_file = False
        kwargs: Dict[str, Any] = dict(file=_file)
        try:
            with open(_file, "r") as f:
                found_file = True
                kwargs.update(yaml.load(f, Loader=Loader))
        except FileNotFoundError:
            try:
                with open(GLOBAL_FILE, "r") as f:
                    kwargs.update(yaml.load(f, Loader=Loader))
            except FileNotFoundError:
                pass

        if load_environment:
            kwargs.update(_from_environment())

        config = cls(**kwargs)

        if not found_file and create_file:
            config.to_file()

        return config

    def to_file(self, file: Optional[str] = None) -> "Config":
        """Save the config to a file."""

        file = file or self.file or default_file()

        with open(file, "w") as f:
            yaml.dump(asdict(self), f, Dumper=Dumper)

        return replace(self, file=file)
