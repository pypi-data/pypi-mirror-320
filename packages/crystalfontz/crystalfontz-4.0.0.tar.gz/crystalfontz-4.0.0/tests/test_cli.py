import json
from typing import Any

import pytest

from crystalfontz.atx import AtxPowerSwitchFunction, AtxPowerSwitchFunctionalitySettings
from crystalfontz.cli import as_json, parse_bytes
from crystalfontz.device import CFA533Status
from crystalfontz.keys import KeyState, KeyStates, KP_UP
from crystalfontz.response import (
    DowDeviceInformation,
    DowTransactionResult,
    GpioRead,
    KeypadPolled,
    LcdMemory,
    Versions,
)


@pytest.mark.parametrize(
    "text,buffer",
    [
        ("hello world", b"hello world"),
        ("\\\\", b"\\"),
        ("\\a", b"\a"),
        ("\\o333", (0o333).to_bytes(1, "big")),
        ("\\o22", (0o22).to_bytes(1, "big")),
        ("\\xff", b"\xff"),
        ("\\", b"\\"),
        ("\\s", b"\\s"),
        ("\\xf", b"\\xf"),
    ],
)
@pytest.mark.filterwarnings("ignore:invalid escape sequence")
def test_parse_bytes(text, buffer) -> None:
    assert parse_bytes(text) == buffer


OBJECTS = [
    Versions(b"CFA533: h1.4, u1v2"),
    LcdMemory(b"\xff\x00\x01\x02\x03\x04\x05\x06\x07"),
    DowDeviceInformation(b"\xff\x00\x01\x02\x03\x04\x05\x06\x07"),
    DowTransactionResult(b"\xff\x01\x02\x03\x04\x05\x06\x07\xff"),
    KeypadPolled(bytes([KP_UP, KP_UP, KP_UP])),
    CFA533Status(
        temperature_sensors_enabled={1, 2},
        key_states=KeyStates(
            up=KeyState(pressed=False, pressed_since=True, released_since=True),
            enter=KeyState(pressed=False, pressed_since=True, released_since=True),
            exit=KeyState(pressed=False, pressed_since=True, released_since=True),
            left=KeyState(pressed=False, pressed_since=True, released_since=True),
            right=KeyState(pressed=False, pressed_since=True, released_since=True),
            down=KeyState(pressed=False, pressed_since=True, released_since=True),
        ),
        atx_power_switch_functionality_settings=AtxPowerSwitchFunctionalitySettings(
            functions={AtxPowerSwitchFunction.KEYPAD_RESET},
            auto_polarity=True,
            reset_invert=False,
            power_invert=False,
            power_pulse_length_seconds=1.0,
        ),
        watchdog_counter=0,
        contrast=0.5,
        keypad_brightness=0.5,
        atx_sense_on_floppy=False,
        cfa633_contrast=0.5,
        lcd_brightness=0.5,
    ),
    GpioRead(bytes([0xFF, 0b0111, 0x11, 0b1101])),
    b"\01",
]


@pytest.mark.parametrize("obj", OBJECTS)
def test_repr(obj: Any, snapshot) -> None:
    assert repr(obj) == snapshot


@pytest.mark.parametrize("obj", OBJECTS)
def test_json(obj: Any, snapshot) -> None:
    assert json.dumps(as_json(obj)) == snapshot
