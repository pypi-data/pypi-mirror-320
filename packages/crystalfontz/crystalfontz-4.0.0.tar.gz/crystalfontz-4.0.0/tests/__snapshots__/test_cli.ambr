# serializer version: 1
# name: test_json[\x01]
  '"AQ=="'
# ---
# name: test_json[obj0]
  '{"model": "CFA533", "hardware_rev": "h1.4", "firmware_rev": "u1v2"}'
# ---
# name: test_json[obj1]
  '{"address": 255, "data": "AAECAwQFBgc="}'
# ---
# name: test_json[obj2]
  '{"index": 255, "rom_id": "AAECAwQFBgc="}'
# ---
# name: test_json[obj3]
  '{"index": 255, "data": "AQIDBAUGBw==", "crc": 255}'
# ---
# name: test_json[obj4]
  '{"states": {"up": {"pressed": true, "pressed_since": true, "released_since": true}, "enter": {"pressed": false, "pressed_since": false, "released_since": false}, "exit": {"pressed": false, "pressed_since": false, "released_since": false}, "left": {"pressed": false, "pressed_since": false, "released_since": false}, "right": {"pressed": false, "pressed_since": false, "released_since": false}, "down": {"pressed": false, "pressed_since": false, "released_since": false}}}'
# ---
# name: test_json[obj5]
  '{"temperature_sensors_enabled": [1, 2], "key_states": {"up": {"pressed": false, "pressed_since": true, "released_since": true}, "enter": {"pressed": false, "pressed_since": true, "released_since": true}, "exit": {"pressed": false, "pressed_since": true, "released_since": true}, "left": {"pressed": false, "pressed_since": true, "released_since": true}, "right": {"pressed": false, "pressed_since": true, "released_since": true}, "down": {"pressed": false, "pressed_since": true, "released_since": true}}, "atx_power_switch_functionality_settings": [{"functions": [32], "auto_polarity": true, "reset_invert": false, "power_invert": false, "power_pulse_length_seconds": 1.0}], "watchdog_counter": 0, "contrast": 0.5, "keypad_brightness": 0.5, "atx_sense_on_floppy": false, "cfa633_contrast": 0.5, "lcd_brightness": 0.5}'
# ---
# name: test_json[obj6]
  '{"index": 255, "state": {"state": true, "falling": true, "rising": true}, "requested_level": 17, "settings": {"function": 8, "mode": 5, "up": "SLOW_STRONG", "down": "SLOW_STRONG"}}'
# ---
# name: test_repr[\x01]
  "b'\\x01'"
# ---
# name: test_repr[obj0]
  'CFA533: h1.4, u1v2'
# ---
# name: test_repr[obj1]
  '0xFF: \\x00\\x01\\x02\\x03\\x04\\x05\\x06\\x07'
# ---
# name: test_repr[obj2]
  '0xFF: \\x00\\x01\\x02\\x03\\x04\\x05\\x06\\x07'
# ---
# name: test_repr[obj3]
  '''
  Transaction Result for Device 255:
    Data: \x01\x02\x03\x04\x05\x06\x07
    CRC: 0xFF
  '''
# ---
# name: test_repr[obj4]
  '''
  Keypad States:
    up: pressed=yes, pressed_since=yes, released_since=yes
    enter: pressed=no, pressed_since=no, released_since=no
    exit: pressed=no, pressed_since=no, released_since=no
    left: pressed=no, pressed_since=no, released_since=no
    right: pressed=no, pressed_since=no, released_since=no
    down: pressed=no, pressed_since=no, released_since=no
  '''
# ---
# name: test_repr[obj5]
  '''
  CFA533 Status:
  --------------
  Temperature sensors enabled: 1, 2
  Key states:
    up: pressed=no, pressed_since=yes, released_since=yes
    enter: pressed=no, pressed_since=yes, released_since=yes
    exit: pressed=no, pressed_since=yes, released_since=yes
    left: pressed=no, pressed_since=yes, released_since=yes
    right: pressed=no, pressed_since=yes, released_since=yes
    down: pressed=no, pressed_since=yes, released_since=yes
  ATX Power Switch Functionality Settings:
    Functions enabled: KEYPAD_RESET
    Auto-Polarity Enabled: yes
    Reset Inverted: no
    Power Inverted: no
    Power Pulse Length (seconds): 1.0
  Watchdog Counter: 0
  Contrast: 0.5
  Contrast (CFA633 Compatible): 0.5
  Backlight:
    Keypad Brightness: 0.5
    LCD Brightness: 0.5
  '''
# ---
# name: test_repr[obj6]
  '''
  GPIO pin 255:
    Requested Level: 17
    Settings:
      Function: 8
      Drive Mode: SLOW_STRONG, SLOW_STRONG
  '''
# ---
